common/include/camera/ICameraControl.hpp

```diff
@@ -0,0 +1,327 @@
+        }
+
+        inline void rotate(const float32_t3& axis, float dRadians)
+        {
+            glm::quat dRotation = glm::angleAxis(dRadians, axis);
+            m_orientation = glm::normalize(dRotation * m_orientation);
+            updateOrthonormalMatrix();
+        }
+
+        inline void strafe(float distance)
+        {
+            move({ 0.f, distance, 0.f });
+        }
+
+        inline void climb(float distance)
+        {
+            move({ 0.f, 0.f, distance });
+        }
+
+        inline void advance(float distance)
+        {
+            move({ distance, 0.f, 0.f });
+        }
+
+        inline void move(float32_t3 delta)
+        {
+            m_position += delta;
+        }
+
+        inline void reset()
+        {
+            // TODO
+        }
+
+        // Position of gimbal
+        inline const float32_t3& getPosition() const { return m_position; }
+
+        // Orientation of gimbal
+        inline const glm::quat& getOrientation() const { return m_orientation; }
+
+        // Orthonormal [getXAxis(), getYAxis(), getZAxis()] matrix
+        inline const float32_t3x3& getOrthonornalMatrix() const { return m_orthonormal; }
+
+        // Base right vector in orthonormal basis, base "right" vector (X-axis)
+        inline const float32_t3& getXAxis() const { return m_orthonormal[0u]; }
+
+        // Base up vector in orthonormal basis, base "up" vector (Y-axis)
+        inline const float32_t3& getYAxis() const { return m_orthonormal[1u]; }
+
+        // Base forward vector in orthonormal basis, base "forward" vector (Z-axis)
+        inline const float32_t3& getZAxis() const { return m_orthonormal[2u]; }
+
+    private:
+        inline void updateOrthonormalMatrix() 
+        { 
+            m_orthonormal = float32_t3x3(glm::mat3_cast(glm::normalize(m_orientation))); 
+
+            // DEBUG
+            const auto [xaxis, yaxis, zaxis] = std::make_tuple(getXAxis(),getYAxis(), getZAxis());
+
+            auto isNormalized = [](const auto& v, float epsilon) -> bool
+            {
+                return glm::epsilonEqual(glm::length(v), 1.0f, epsilon);
+            };
+
+            auto isOrthogonal = [](const auto& a, const auto& b, float epsilon) -> bool
+            {
+                return glm::epsilonEqual(glm::dot(a, b), 0.0f, epsilon);
+            };
+
+            auto isOrthoBase = [&](const auto& x, const auto& y, const auto& z, float epsilon = 1e-6f) -> bool
+            {
+                return isNormalized(x, epsilon) && isNormalized(y, epsilon) && isNormalized(z, epsilon) &&
+                    isOrthogonal(x, y, epsilon) && isOrthogonal(x, z, epsilon) && isOrthogonal(y, z, epsilon);
+            };
+
+            assert(isOrthoBase(xaxis, yaxis, zaxis));
+        }
+
+        float32_t3 m_position;
+        glm::quat m_orientation;
+
+        // Represents the camera's orthonormal basis
+        // https://en.wikipedia.org/wiki/Orthonormal_basis
+        float32_t3x3 m_orthonormal;
```

Comment:
only keep the quaternion, and spit out the matrix on demand, its not that expensive to compute it

Reply:
Also keep everything as `float64_t`